server/cluster/cluster.go

```diff
@@ -1252,6 +1300,115 @@ func (c *RaftCluster) checkStores() {
 	}
 }
 
+func (c *RaftCluster) getThreshold(stores []*core.StoreInfo, store *core.StoreInfo, keys [][]byte) float64 {
+	storeSize := 0.0
+	if len(keys) == 0 {
+		return c.calculateRange(stores, store, []byte(""), []byte("")) * 0.9
+	}
+	startKey := []byte("")
+	for _, key := range keys {
+		endKey := key
+		storeSize += c.calculateRange(stores, store, startKey, endKey)
+		startKey = endKey
+	}
+
+	return storeSize * 0.9
+}
+
+func (c *RaftCluster) calculateRange(stores []*core.StoreInfo, store *core.StoreInfo, startKey, endKey []byte) float64 {
+	var storeSize float64
+	rules := c.ruleManager.GetRulesForApplyRange(startKey, endKey)
+	for _, rule := range rules {
+		if !placement.MatchLabelConstraints(store, rule.LabelConstraints) {
+			continue
+		}
+
+		var matchStores []*core.StoreInfo
+		for _, store := range stores {
+			if store.IsRemoving() || store.IsRemoved() {
+				continue
+			}
+			if placement.MatchLabelConstraints(store, rule.LabelConstraints) {
+				matchStores = append(matchStores, store)
+			}
+		}
+		regionSize := c.core.GetRegionSizeByRange(rule.StartKey, rule.EndKey) * int64(rule.Count)
+		weight := getStoreTopoWeight(store, matchStores, rule.LocationLabels)
+		storeSize += float64(regionSize) * weight
+		log.Debug("calculate range result",
+			zap.Uint64("store-id", store.GetID()),
+			zap.String("rule", rule.String()),
+			zap.Int64("region-size", regionSize),
+			zap.Float64("weight", weight),
+			zap.Float64("store-size", storeSize),
+		)
+	}
+	return storeSize
+}
+
+func getStoreTopoWeight(store *core.StoreInfo, stores []*core.StoreInfo, locationLabels []string) float64 {
```

Comment:
If the isolation level does not meet the number of replica, there seems to be some offset.